# HG changeset patch
# User Betterbird <betterbird@betterbird.eu>
# Date 1785847252 -7200
# Parent  4aa3afe11227ddd187b36d787cb668d1bf9d8578
Feature: Show most recent child in collapsed thread root, fix integration of newly arrived message.

Also fix intial view in XF views by fixing the thread roots in nsMsgXFVirtualFolderDBView::OnSearchDone().

diff --git a/mailnews/base/src/nsMsgThreadedDBView.cpp b/mailnews/base/src/nsMsgThreadedDBView.cpp
--- a/mailnews/base/src/nsMsgThreadedDBView.cpp
+++ b/mailnews/base/src/nsMsgThreadedDBView.cpp
@@ -589,9 +589,10 @@ nsresult nsMsgThreadedDBView::OnNewHeade
   bool isExpanded = !(flags & nsMsgMessageFlags::Elided);
 
   // Return early if the message won't be visible in the view and won't change
   // the order of the threads.
-  if (!isExpanded && aParentKey != nsMsgKey_None && !moveThread) {
+  if (!isExpanded && aParentKey != nsMsgKey_None && !moveThread &&
+      !CollapseToMostRecent()) {
     // Update the parent thread's unread and total counts.
     NoteChange(threadIndex, 1, nsMsgViewNotificationCode::changed);
     return NS_OK;
   }
@@ -637,9 +638,9 @@ nsresult nsMsgThreadedDBView::OnNewHeade
   if (isExpanded) {
     CollapseByIndex(threadIndex, nullptr, false);
   }
 
-  if (ReversedThreading()) {
+  if (isExpanded && ReversedThreading()) {
     // Correct `threadIndex`, since all the children above
     // are no longer displayed.
     // Since the new header is already counted in the thread, but not displayed
     // yet, we need to subtract it. Also, collapsing a thread with N children,
@@ -668,14 +669,34 @@ nsresult nsMsgThreadedDBView::OnNewHeade
     nsMsgDBView::AddHdr(newHdr, &newIndex);
   } else if (moveThread) {
     // Chronological shift: move the root to its new sorted position.
     nsCOMPtr<nsIMsgDBHdr> existingRoot;
-    rv = nsMsgDBView::GetMsgHdrForViewIndex(threadIndex,
-                                            getter_AddRefs(existingRoot));
-    NS_ENSURE_SUCCESS(rv, rv);
+    if (isExpanded || !CollapseToMostRecent()) {
+      rv = nsMsgDBView::GetMsgHdrForViewIndex(threadIndex,
+                                              getter_AddRefs(existingRoot));
+      NS_ENSURE_SUCCESS(rv, rv);
+    }
     rv = nsMsgDBView::RemoveByIndex(threadIndex);
     NS_ENSURE_SUCCESS(rv, rv);
-    nsMsgDBView::AddHdr(existingRoot, &newIndex);
+    if (!isExpanded && CollapseToMostRecent()) {
+      nsMsgDBView::AddHdr(newHdr, &newIndex);
+    } else {
+      nsMsgDBView::AddHdr(existingRoot, &newIndex);
+    }
+  } else if (!isExpanded && CollapseToMostRecent()) {
+    uint32_t newestMsgDate = 0, newMsgDate = 0;
+    threadHdr->GetNewestMsgDate(&newestMsgDate);
+    newHdr->GetDateInSeconds(&newMsgDate);
+    if (newMsgDate == newestMsgDate) {
+      uint32_t childFlags;
+      newHdr->GetFlags(&childFlags);
+      m_keys[threadIndex] = newKey;
+      m_flags[threadIndex] =
+          childFlags | nsMsgMessageFlags::Elided |
+          (m_flags[threadIndex] &
+           (MSG_VIEW_FLAG_ISTHREAD | MSG_VIEW_FLAG_HASCHILDREN));
+      SetFolder(threadIndex, newHdr);
+    }
   }
 
   if (newIndex != nsMsgViewIndex_None) {
     targetIndex = newIndex;
diff --git a/mailnews/base/src/nsMsgXFVirtualFolderDBView.cpp b/mailnews/base/src/nsMsgXFVirtualFolderDBView.cpp
--- a/mailnews/base/src/nsMsgXFVirtualFolderDBView.cpp
+++ b/mailnews/base/src/nsMsgXFVirtualFolderDBView.cpp
@@ -357,8 +357,49 @@ nsMsgXFVirtualFolderDBView::OnSearchDone
     m_viewFolder->UpdateSummaryTotals(true);
     virtDatabase->Commit(nsMsgDBCommitType::kLargeCommit);
   }
 
+  if (CollapseToMostRecent()) {
+    for (uint32_t i = 0; i < m_flags.Length(); i++) {
+      if (!(m_flags[i] & nsMsgMessageFlags::Elided)) continue;
+      if (!(m_flags[i] & MSG_VIEW_FLAG_HASCHILDREN)) continue;
+
+      nsCOMPtr<nsIMsgThread> thread;
+      GetThreadContainingIndex(i, getter_AddRefs(thread));
+      if (!thread) continue;
+
+      uint32_t numChildren;
+      thread->GetNumChildren(&numChildren);
+
+      uint32_t newestDate = 0;
+      nsMsgKey newestKey = nsMsgKey_None;
+      nsCOMPtr<nsIMsgDBHdr> newestHdr;
+      for (uint32_t childIndex = 0; childIndex < numChildren; childIndex++) {
+        nsCOMPtr<nsIMsgDBHdr> child;
+        thread->GetChildHdrAt(childIndex, getter_AddRefs(child));
+        if (!child) continue;
+        uint32_t msgDate;
+        child->GetDateInSeconds(&msgDate);
+        if (msgDate > newestDate) {
+          newestDate = msgDate;
+          child->GetMessageKey(&newestKey);
+          newestHdr = child;
+        }
+      }
+
+      if (newestKey != nsMsgKey_None && newestKey != m_keys[i]) {
+        uint32_t childFlags;
+        newestHdr->GetFlags(&childFlags);
+        m_keys[i] = newestKey;
+        nsCOMPtr<nsIMsgFolder> msgFolder;
+        newestHdr->GetFolder(getter_AddRefs(msgFolder));
+        m_folders.ReplaceObjectAt(msgFolder, i);
+        m_flags[i] = childFlags | nsMsgMessageFlags::Elided |
+            (m_flags[i] & (MSG_VIEW_FLAG_ISTHREAD | MSG_VIEW_FLAG_HASCHILDREN));
+      }
+    }
+  }
+
   if (!m_sortValid && m_sortType != nsMsgViewSortType::byThread &&
       !(m_viewFlags & nsMsgViewFlagsType::kThreadedDisplay)) {
     // Sort the results.
     m_sortValid = false;
